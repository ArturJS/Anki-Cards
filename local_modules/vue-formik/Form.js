import { createForm, formSubscriptionItems } from 'final-form';
import { getChildren } from './utils';

const defaultSubscription = formSubscriptionItems.reduce((result, key) => {
  result[key] = true;

  return result;
}, {});

const createValidatorFromSchema = validationSchema => {
  return async values => {
    try {
      await validationSchema.validate(values, {
        abortEarly: false
      });
    } catch (error) {
      const errors = error.inner.reduce(
        (collectedErrors, { path, message }) => {
          return {
            ...collectedErrors,
            [path]: message
          };
        },
        {}
      );

      return errors;
    }
  };
};

export default {
  name: 'final-form',

  props: {
    initialValues: Object,
    submit: {
      type: Function,
      default: () => {}
    },
    subscription: Object,
    validationSchema: Object
  },

  provide() {
    return {
      finalForm: this.finalForm
    };
  },

  data() {
    return {
      finalForm: createForm({
        onSubmit: this.submit,
        initialValues: this.initialValues,
        validate: this.getValidator()
      }),
      formState: null
    };
  },

  methods: {
    getValidator() {
      return createValidatorFromSchema(this.validationSchema);
    },
    handleSubmit(e) {
      e && e.preventDefault();
      this.finalForm.submit();
    }
  },

  created() {
    this.unsubscribe = this.finalForm.subscribe(state => {
      this.formState = state;
      this.$emit('change', state);
    }, this.subscription || defaultSubscription);
  },

  beforeDestroy() {
    this.unsubscribe();
  },

  render(h) {
    return (
      <form class={['vf-form', this.class]} onSubmit={this.handleSubmit}>
        {getChildren(this.$slots.default)}
      </form>
    );
  }
};
