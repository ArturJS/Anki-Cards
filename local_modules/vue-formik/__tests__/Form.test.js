import { shallowMount } from '@vue/test-utils';
import Form from '../Form';

jest.mock('../utils', () => ({
  getChildren: children => `getChildren( ${JSON.stringify(children, null, 2)} )`
}));

describe('Check <Form/>', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(Form, {
      slots: {
        default: 'Test children'
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
