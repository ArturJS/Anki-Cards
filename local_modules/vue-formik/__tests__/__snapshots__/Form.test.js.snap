// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check <Form/> renders correctly 1`] = `
<form class="vf-form">getChildren( [
  {
  "text": "Test children",
  "raw": false,
  "isStatic": false,
  "isRootInsert": true,
  "isComment": false,
  "isCloned": false,
  "isOnce": false,
  "isAsyncPlaceholder": false
  }
  ] )</form>
`;
